src/buildercore/core.py

```diff
@@ -246,12 +246,16 @@ def stack_pem(stackname, die_if_exists=False, die_if_doesnt_exist=False):
 
 def _ec2_connection_params(stackname, username, **kwargs):
     "returns a dictionary of settings to be used with Fabric's api.settings context manager"
+    # http://docs.fabfile.org/en/1.14/usage/env.html
     params = {'user': username}
     pem = stack_pem(stackname)
     # handles cases where we want to establish a connection to run a task
     # when machine has failed to provision correctly.
-    if os.path.exists(pem) and username == config.BOOTSTRAP_USER:
-        params['key_filename'] = pem
+    if username == config.BOOTSTRAP_USER:
+        if os.path.exists(pem):
+            params['key_filename'] = pem
+        else:
+            raise RuntimeError("private key for the bootstrap user for this host does not exist. I looked here: %s" % pem)
```

Comment:
I think that then this doesn't cover the use case of using your own `.ssh/id_rsa` key rather than the one in `builder`. If you are not the user that created the stack, probably won't have that one. Or should it be synced from S3?

Reply:
In the case of https://alfred.elifesciences.org/blue/organizations/jenkins/pull-requests-projects%2Fapi-dummy/detail/PR-152/1/pipeline it's definitely on S3 so there is a missing auto-download step.

Reply:
> Or should it be synced from S3?
>  it's definitely on S3 so there is a missing auto-download step.

yup, it should be automatically downloaded from S3. It's making that particular failure explicit. I encountered it a handful of times while doing the upgrades.